### DISCUSSION TITLE
Get path template URL in middleware

### DISCUSSION THREAD

**Discussion by dan-cbm** [fastapi/fastapi#12975] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class HTTPMiddleware(BaseHTTPMiddleware):
    def __init__(self, app):
        super().__init__(app)

    async def dispatch(self, request: Request, call_next):
        response = await call_next(request)
        # print path like  /api/example/1234/cheese/567/info
        logger.info(f"Received {request.method} on {request.url.path}")

        # TODO: print path as well like  /api/example/{company_id}/cheese/{site_id}/info 
        return response
```

### Description

I'm working on some FastAPI middlewear which we use for logging all of our API requests. At present I'm logging the request path, but I'd like to also log the template URL for the endpoint e.g. the literal string `/api/example/{company_id}/cheese/{site_id}/info` so that we can later filter our logs and see all request to a given endpoint, regardless of the particular IDs in the url.

Is this possible with FastAPI?  Somehow Sentry is able to tag errors with the endpoint URL in the "template form" we want, but I can't figure out how to access that information myself.

### Operating System

Linux

### FastAPI Version

0.104.1

### Pydantic Version

2.6.1

### Python Version

3.11.10

**Comment by luzzodev** [comment, marked as answer]
The routing phase will inject route info in request, so you can can access to those information after you received your response.

In your case:

``` python
class HTTPMiddleware(BaseHTTPMiddleware):
    def __init__(self, app):
        super().__init__(app)

    async def dispatch(self, request: Request, call_next):
        response = await call_next(request)
        # print path like  /api/example/1234/cheese/567/info
        print(f"Received {request.method} on {request.url.path}")

        # TODO: print path as well like  /api/example/{company_id}/cheese/{site_id}/info
        print(request.scope["route"].path)
        return response
```

Just be careful that if a response fails, exceptions are not handled, or the routing phase is never reached due to an error, then the routing information will not be injected. Always check if the route key exists.
